amazon-kinesis-client/src/main/java/software/amazon/kinesis/metrics/CloudWatchMetricsPublisher.java

```diff
@@ -60,11 +62,11 @@ public void publishMetrics(List<MetricDatumWithKey<CloudWatchMetricKey>> dataToP
             request = request.metricData(metricData);
 
             try {
-                cloudWatchClient.putMetricData(request.build());
-
-                log.debug("Successfully published {} datums.", endIndex - startIndex);
-            } catch (CloudWatchException e) {
-                log.warn("Could not publish {} datums to CloudWatch", endIndex - startIndex, e);
+                // This needs to be blocking. Making it asynchronous leads to increased throttling.
+                cloudWatchClient.putMetricData(request.build()).get(PUT_TIMEOUT_MILLIS, TimeUnit.MILLISECONDS);
```

Comment:
No need to add the timeout. Could you catch for specific exceptions instead of generic ones.

```java
final AWSExceptionManager exceptionManager = new AWSExceptionManager();
exceptionManager.add(CloudWatchException.class, t -> t);
try {
    try {
        cloudWatchClient.putMetricData(request.build()).get();
    } catch (ExecutionException e) {
        throw exceptionManager.apply(e.getCause);
    } catch (InterruptedException e) {
        log.info("Interrupted, shutdown invoked");
    }
}
catch(CloudWatchException e) {
    log.warn("Exception thrown while publishing to CW", e);
} catch (Exception e) {
    log.error("Unknown issue while publishing to CW", e);
}
```

Reply:
Synced up offline. We will go ahead with timeouts.